GRDB/Core/FetchRequest.swift

```diff
@@ -15,10 +15,20 @@ public protocol FetchRequest: DatabaseRegionConvertible {
     /// Returns a tuple that contains a prepared statement that is ready to be
     /// executed, and an eventual row adapter.
     ///
+    /// Default implementation uses `prepare(db, hint: nil)`.
```

Comment:
Second question: what is the purpose of keeping this method in the protocol? Is there any reason for a type to implement both methods? Isn't `prepare(_:hint:)` sufficient?

Reply:
Because the protocol's `public` I had this idea of not wanting to break compatibility with anything already calling `prepare(_:)`. Also thought `hint` could be confusing when you don't need it.

I'd have preferred to declare it as:

```
func prepare(_ db: Database, hint: FetchRequestHint? = nil) throws -> (SelectStatement, RowAdapter?)
```

… but you can't have a default argument in a protocol.

Anyway, I'm not too attached to it. Happy to take it out if you'd prefer.

Reply:
All right, I understand. The good news is that we can break things, because we are targeting the next major version. GRDB 4 will break several things (see #479 for the list of changes). So we can keep a single method.

I agree the hint adds a layer of complexity to the API, but I don't expect many users to write their own concrete type which conforms to FetchRequest. This is an advanced technique. If they do, I guess they'll be able to handle it.

But we have to document clearly that this hint is only an opportunity for optimization, and can safely be ignored.

Towards this goal of clarity, I think we'll also have to look for alternative names for `hint` and `FetchRequestHint`, and wonder if we really want an optional. Those currently sound too much like an implementation detail, and not enough like a public API.

I don't know... The radical solution is:

```swift
func prepare(_ db: Database, forSingleResult singleResult: Bool) -> ... {
    if singleResult {
        // Add LIMIT 1
        return ...
    } else {
        // Don't add LIMIT 1
        return ...
    }
}
```
 
But a boolean has little room for further extensions (even if I can't foresee any right now). I agree with your idea of an enum:

```swift
func prepare(_ db: Database, forCardinality cardinality: FetchCardinality) -> ... {
    switch cardinality {
    case .single:
        // Add LIMIT 1
        return ...
    case .multiple:
        // Don't add LIMIT 1
        return ...
    }
}
```

But "cardinality" can't really be extended either. Not really an improvement over a boolean. Let's go all general:

```swift
func prepare(_ db: Database, forPurpose purpose: FetchPurpose) -> ... {
    switch purpose {
    case .singleResult:
        // Add LIMIT 1
        return ...
    case .multipleResults:
        // Don't add LIMIT 1
        return ...
    }
}
```

I wish we could have merged `FetchRequest.fetchCount` here:

```swift
func prepare(_ db: Database, forPurpose purpose: FetchPurpose) -> ... {
    switch purpose {
    case .singleResult:
        // Add LIMIT 1
        return ...
    case .multipleResults:
        // Don't add LIMIT 1
        return ...
    case .count:
        // Build the counting statement
        return ...
    }
}
```

But we can't choose this last route, because `fetchCount` is currently an optional method with a non-trivial default implementation: users would now have real difficulties adopting the FetchRequest protocol.

Well.. As you see, we have much room in our way to the best design. I'm curious about your ideas as well. Our goal: turn our implementation strategy into a public API that can make sense for people who **don't have a clue** about our implementation strategy. I mean that most users are utterly indifferent about how the library works, and only care about their needs. Good public APIs acknowledge this hard fact. It may look difficult, but this can also be a funny game :-)

A technique that has constantly helped me finding an answer to this kind of question has been the README. We currently write, below the introduction to the FetchRequest protocol:

> The `prepare` method returns a prepared statement and an optional row adapter. The [prepared statement](#prepared-statements) tells which SQL query should be executed. The row adapter helps presenting the fetched rows in the way expected by the row decoders (see [row adapter](#row-adapters)).

We can try to update this paragraph so that it explains how to use the optional hint. Experience shows that when the documentation looks like it fills its teaching goal, the API is good as well.

Reply:
Applying my own advice makes the plain and simple boolean approach look not so bad:

> The `prepare(_:forSingleResult:)` method accepts a database connection, a `singleResult` hint, and returns a prepared statement and an optional row adapter. The `singleResult` hint can safely be ignored, but conforming types can use it as an opportunity to append `LIMIT 1` to the returned [prepared statement](#prepared-statements). The optional row adapter helps presenting the fetched rows in the way expected by the row decoders (see [row adapter](row-adapters)).

Reply:
I'm coming around to the idea of a plain old boolean. Complicating the API for unknown future cases probably isn't worth it. If other scenarios do come up in the future, it can be changed in a way that really makes sense.

Reply:
> Complicating the API for unknown future cases probably isn't worth it.

Indeed, if we really can't find any extension for the "hint" today, then let's not try to outsmart ourselves. As you say, we'll always be able to revise our opinion in the future, if needed!

Reply:
Cool. I've switched it to a boolean and updated the README.